base/file.jl

```diff
@@ -510,22 +510,47 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
     return (filename, Base.open(filename, "r+"))
 end
 
+let
+global tempname
+
+struct GUID
+    l1::Culong
+    w1::Cushort
+    w2::Cushort
+    b::NTuple{8,Cuchar}
+end
+
 function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
     isdir(parent) || throw(ArgumentError("$(repr(parent)) is not a directory"))
-    seed::UInt32 = rand(UInt32)
-    while true
-        if (seed & typemax(UInt16)) == 0
-            seed += 1
-        end
-        filename = _win_tempname(parent, seed)
-        if !ispath(filename)
-            cleanup && temp_cleanup_later(filename)
-            return filename
-        end
-        seed += 1
+
+    RPC_S_UUID_LOCAL_ONLY = 1824
+    RPC_S_UUID_NO_ADDRESS = 1739
+    RPC_S_OUT_OF_MEMORY = 14
+
+    id = Ref{GUID}()
+    r = ccall((:UuidCreate,:Rpcrt4), Cint, (Ref{GUID},), id)
+    r == 0 || (r == RPC_S_UUID_LOCAL_ONLY && (@warn "UIUD guaranteed to be unique to this computer only"; true)) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))
+
+    nameptr = Ref{Ptr{Cwchar_t}}()
+    r = ccall((:UuidToStringW, :Rpcrt4), Cint, (Ref{GUID}, Ref{Ptr{Cwchar_t}}), id, nameptr)
+    r == 0 || (r == RPC_S_OUT_OF_MEMORY && error("Out of memory")) || error("cannot convert UIUD to string")
+
+    namebuf = unsafe_wrap(Vector{Cwchar_t}, nameptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), nameptr[]))
+    name = transcode(String, namebuf)
+
+    @assert ccall((:RpcStringFreeW, :Rpcrt4), Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, name)
+    if !ispath(filename)
```

Comment:
Are we ok with this. I feel like this should just error (maybe?), since it will be an incredible rare scenario. In the previous code, we did keep looping until a unique name was generated. Here,  in the unlikely scenario it is a path, we recursively call the function.

Reply:
I think maybe if `cleanup` is false we could not call `ispath` at all? It would be nice if this function would actually not touch the file system at all.

I tend to think that just assuming we'll never run into a conflict is ok here?

Reply:
> I think maybe if cleanup is false we could not call ispath at all?

Agreed, in fact it's not guaranteed to return a unique filename (even if very unlikely)

Reply:
Since @StefanKarpinski worked on this, could you comment on  ?

> It would be nice if this function would actually not touch the file system at all.

I don't see how it would be possible to do this in the scenario, since if we don't check `!ispath(filename)` ,   `temp_cleanup_later` could accidentally remove the file which we did not generate.